### PR TITLE
Add function to set a nested field

### PR DESCRIPTION
**Pull request by charignon**
Add support for nested fields.

#### Test Plan
- [X] Added a unit test

### CODE REVIEW

**Comment by dotemacs** [comment]
Thank you for doing this.

Let me ask you about the reasoning behind this:

why create this as a separate function, instead of extending the `set-fields` that would handle the nested fields?

Thanks

**Comment by charignon** [comment]
@dotemacs what do you think of the new version I pushed?

**Comment by dotemacs** [comment]
Hey @charignon 

I actually prefer this version over the old one :)

One thing that I'd do is maybe reword the doc string like this:

```clojure
"Fill in the form fields with the `new-fileds`.
If `new-fields` is a map {field value ...} fill in the top-level fields listed in the map.
But if it's a vector of [[value index derivedName] ...] fill in the nested fields."
```

The above is good if we're to stick to the current version.

I'd like to ask you for some further clarification, because I'm neck deep into something else...

What I'm trying to understand is if we have children of a particular field, with your solution, the replacement would happen on the numeric position of the child, fore example using the test:

```clojure
[["25" 0 "Amount"]]
```

That is saying for the child number 0 (the first child), find the field named "Amount" and replace it with "25", right ?

If the answer to my question above is yes, what do you think of this syntax instead:

```clojure
[[0 "Amount" "25" ]]
```

The reason I ask this is because I'm following this logic: find the child (here it's 0), then find the filed within that child (here that's "Amount") and add the value (25).

Now for the next question:

But let's say that you like that syntax, how would a mix of the two work?

Say that there was a form that had 

```
{"foo" "bar" "quux" ["x" "y" "z" "p"]}
```

(The field `quux` here being the field with the children.)

Would the update for `z` look like:

```clojure
{"quux" [[1 "z" "new-value"]]}
```
Where the value `p` would be replaced with `new-value`.

**Comment by dotemacs** [comment]
Looked at this a bit at the REPL...

If you get the list of the fields in the PDF you've added `claim.pdf`.
In it, you'll find a field named `Amount`. 
Then if you look at the children of it, you'll see the following names for them:

```clojure
pdfboxing.form> (map #(.getFullyQualifiedName %) (.getChildren amount-field))
("Amount.0" "Amount.1" "Amount.2")
```
Which to means that the function `get-fields` should be extended to not only look at the top level fields of the document, but to check for their children. And when it does that, it should return them all. Something like:

```clojure
{...
"Amount": {"Amount.0" "" "Amount.1" "" "Amount.2" ""}
...}
```

Then the syntax for the update of the values can be:

```clojure
(set-fields {"Amount" {"Amount.0" "25"}})
```

This way it's more "natural", as in idiomatic Clojure. Because you'll be working with "just data" rather than having to figure out the numerical position of the field you'd want to update.

What do you think about this?

I'll tackle the rework of `get-fields` so that it returns all the fields.

**Comment by dotemacs** [comment]
Hey @charignon I created this PR: https://github.com/charignon/pdfboxing/pull/1 

As mentioned previously, the idea was to really get **all the fields** from a PDF document, rather than just the top level ones.

**Comment by dotemacs** [comment]
Looked into this a bit more.

And it seems that the code change might not be needed after all to `set-fields`.

See (this is after the update to `get-fields`):

```clojure
pdfboxing.form> (get-fields "./test/pdfs/claim.pdf")
{...
 "Amount" {"Amount.0" "", "Amount.1" "", "Amount.2" ""},
...}
```

If you then do:

```clojure
pdfboxing.form> (set-fields "./test/pdfs/claim.pdf" "/tmp/output.pdf" {"Amount.0" "ninja"})
nil
pdfboxing.form> (get-fields "/tmp/output.pdf")
{...
 "Amount" {"Amount.0" "ninja", "Amount.1" "", "Amount.2" ""},
...}
```

The nested field gets filled in. Because the name of the field is
unique, you don't need to use the vector or to use a nested map to
provide the value.


Intrestingly if you try to fill in the top level field, then it fills
in all the child fields with the given value:

```clojure
pdfboxing.form> (set-fields "./test/pdfs/claim.pdf" "/tmp/output.pdf" {"Amount" "ninja"})
nil
pdfboxing.form> (get-fields "/tmp/output.pdf")
{...
 "Amount"
 {"Amount.0" "ninja", "Amount.1" "ninja", "Amount.2" "ninja"},
...}
```

I think that you prompted an interesting point with your PR because it
prompted me to improve the library.

**Comment by dotemacs** [comment]
After a bit more tweaking, I've created this: [nested-field-work](https://github.com/dotemacs/pdfboxing/tree/nested-field-work) branch, where I've added all the work that was discussed above.

I've specifically mention you in this commit (which is also left with your name, to honour your contribution): https://github.com/dotemacs/pdfboxing/commit/c26761188131f0918189872b3903934474a568ee

Let me know what you think, thanks.

**Comment by charignon** [comment]
@dotemacs I looked at your change. I understand your point about working with "just data" and it makes it easier to literally map a hash to a form directly, let's go with your branch :) 
Thanks for all your work on this!

**Comment by charignon** [comment]
If you are fine then, let's close this PR and merge your change in your repo and release a new version.

**Comment by dotemacs** [comment]
> @dotemacs I looked at your change. I understand your point about working with "just data" and it makes it easier to literally map a hash to a form directly, let's go with your branch :)
> Thanks for all your work on this!

I think you've nudged me in that direction, so thanks again :)
I'm glad you like it.

> If you are fine then, let's close this PR and merge your change in your repo and release a new version.

I just cut: [0.1.15-SNAPSHOT](https://clojars.org/pdfboxing/versions/0.1.15-SNAPSHOT).

Thanks
